### ISSUE TITLE
Performance issue with TemplateContext.CacheMemory

### ISSUE THREAD

**Issue by marko444** [ServiceStack/Issues#653]
Rendering "each" inside a template in a loop 10000 times:

```hbs
{{#each document in Model.Links}}
    <tr>
        <td class="msgText" align="left" valign="center">
            <a href="{{document.Link}}" style="text-decoration:none;">
                <span style="color: #DB4453;font-size:12px;">
                    {{document.Name}}
                </span>
            </a>
        </td>
    </tr>
{{/each}}
```

Is adding 10000 items "#each document in Model.Links" to the CacheMemory. 

`public ConcurrentDictionary<ReadOnlyMemory<char>, object> CacheMemory { get; }`

**Comment by mythz** [comment]
Please provide something I can run, where is the loop?

**Comment by marko444** [comment]
```
        [Fact]
        public void UniqueTemplateShouldBeCachedOnlyOnce()
        {
            var context = new TemplateContext
            {
                TemplateBlocks = { new TemplateEvalBlock() },
                Args = {
                    ["templates"] = new List<string> {
                        "1. {{income ?? 1000}} - {{expenses}}",
                        "2. {{income ?? 2000}} - {{expenses}}",
                        "3. {{income ?? 3000}} - {{expenses}}",
                    }
                }
            }.Init();

            10000.Times(() =>
            {
                var result = context.EvaluateTemplate(@"{{#each templates}}{{index}}{{/each}}");
                Assert.Equal("012", result);
            });

            // Fails with count = 10000
            Assert.True(context.CacheMemory.Count == 1);
        }
```

**Comment by mythz** [comment]
You're seeing this issue because you're using `context.EvaluateTemplate()` in the loop which creates a new "dynamic page" behind-the-scenes where a `ReadOnlyMemory<char>` of the same string in one page doesn't equal the same string in a different page (which IMO is unexpected).

This is less of an issue in practice as the same page is cached and used so the keys will match what's in the cache. E.g. it will use the same cache entry if you use the same page:

```csharp
var page = context.OneTimePage(@"{{#each templates}}{{index}}{{/each}}");
10000.Times(() =>
{
    var result = new PageResult(page).Result;
});
```

But as it unexpectedly ignores the cache when the same code fragment is used in multiple pages I've switched it to use the string key Cache in [this commit](https://github.com/ServiceStack/ServiceStack/commit/51c8272b77ae9cd1cbe23fe4a4a90ce1a5caac0c).

This change is available in the latest v5.4.1 that's now [available on MyGet](https://docs.servicestack.net/myget).
